File: elixir/2015/day01b/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        result =
          content
          |> String.graphemes()
          |> Enum.reduce_while({0, 0}, fn
            "(", {pos, floors} ->
              floors = floors + 1
              {:cont, {pos + 1, floors}}
            ")", {pos, floors} ->
              floors = floors - 1
              if floors < 0 do
                {:halt, {:found, pos + 1}}
              else
                {:cont, {pos + 1, floors}}
              end
            _, {pos, floors} ->
              {:cont, {pos + 1, floors}}
          end)
        case result do
          {:found, pos} -> {:ok, pos}
          _ -> {:error, "result was never negative"}
        end
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: erlang/2015/day01b/Program.erl
#!/usr/bin/env escript

main(Args) ->
  case Args of
    [Filename] ->
      case process(Filename) of
        {ok, Result} ->
          io:format("result = ~p~n", [Result]);
        {error, Reason} ->
          io:format(standard_error, "error while processing file `~s': ~p~n", [Filename, Reason]),
          halt(1)
      end;
    _ ->
      usage()
  end.

usage() ->
  io:format(standard_error, "usage: Program.erl <input file>~n", []),
  halt(1).

process(Filename) ->
  case file:read_file(Filename) of
    {ok, Chars} ->
      Content = unicode:characters_to_list(Chars),
      case find_basement(Content, 1, 0) of
        {found, Pos} -> {ok, Pos};
        not_found -> {error, "result was never negative"}
      end;
    {error, Reason} ->
      {error, Reason}
  end.

find_basement([], _, _) ->
  not_found;
find_basement([$( | Rest], Pos, Floor) ->
  find_basement(Rest, Pos + 1, Floor + 1);
find_basement([$) | Rest], Pos, Floor) ->
  NewFloor = Floor - 1,
  if
    NewFloor < 0 -> {found, Pos};
    true -> find_basement(Rest, Pos + 1, NewFloor)
  end;
find_basement([_ | Rest], Pos, Floor) ->
  find_basement(Rest, Pos + 1, Floor).
